app/com/arpnetworking/metrics/portal/alerts/impl/CachingAlertExecutionRepository.java

```diff
@@ -0,0 +1,243 @@
+                    }
+                    return _inner.getLastSuccess(jobId, organization)
+                        .thenCompose(res2 -> {
+                            if (!res2.isPresent()) {
+                                return CompletableFuture.completedFuture(res2);
+                            }
+                            return AlertExecutionCacheActor.put(
+                                    _successCache,
+                                    organization,
+                                    res2.get(),
+                                    _cacheOperationTimeout
+                            ).thenApply(ignore -> res2);
+                        });
+                });
+    }
+
+    @Override
+    public CompletionStage<ImmutableMap<UUID, JobExecution.Success<AlertEvaluationResult>>> getLastSuccessBatch(
+        final List<UUID> jobIds,
+        final Organization organization,
+        final LocalDate maxLookback
+    ) throws NoSuchElementException {
+        // Attempt to get all ids from cache
+        final CompletionStage<ImmutableMap<UUID, JobExecution.Success<AlertEvaluationResult>>> cached =
+                AlertExecutionCacheActor.multiget(
+                    _successCache,
+                    organization,
+                    jobIds,
+                    _cacheOperationTimeout
+                );
+
+        return cached.thenCompose(hits -> {
+            // Check for any cache misses and fetch those from the inner
+            // repository.
+            final List<UUID> misses =
+                jobIds.stream()
+                    .filter(Predicates.in(hits.keySet()).negate())
+                    .collect(ImmutableList.toImmutableList());
+            if (misses.isEmpty()) {
+                return CompletableFuture.completedFuture(hits);
+            }
+            return _inner
+                .getLastSuccessBatch(misses, organization, maxLookback)
+                .whenComplete((rest, error) -> {
+                    if (error != null) {
+                        return;
+                    }
+                    writeBatchToCache(rest, organization);
+                })
+                .thenApply(rest -> {
+                    // Merge the cache hits / misses into a single map.
+                    return Stream.concat(
+                        hits.entrySet().stream(),
+                        rest.entrySet().stream()
+                    ).collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+                });
+        });
+    }
+
+    private void writeBatchToCache(
+        final Map<UUID, JobExecution.Success<AlertEvaluationResult>> batch,
+        final Organization organization
+    ) {
+        batch.values()
+            .forEach(e -> {
+//                final String key = cacheKey(e.getJobId(), organization.getId());
+                AlertExecutionCacheActor.put(_successCache, organization, e, _cacheOperationTimeout);
```

Comment:
I'm confused, isn't a batch of N items still N messages to the actor?

Reply:
You're right, I batched the get but not the fallback put. Addressed.